### PR TITLE
Allow `#[derive(GraphQLInterface)]` and `#[graphql_interface]` on structs

### PR DESCRIPTION
**Pull request by ilslv**
## Synopsis

After #1009 GraphQL interfaces are represented with traits acting as blueprints for methods names, arguments and return types.

## Solution

To emphasise structural nature of GraphQL interfaces, allow representing them with structs:

```rust
#[derive(GraphQLInterface)]
#[graphql(for = Human)]
struct Character {
    id: String,
}
```

```rust
#[graphql_interface(for = Human)]
struct Character {
    id: String,
}
```

## Checklist

- Created PR:
    - [x] In [draft mode][l:1]
    - [x] Name contains `Draft: ` prefix
    - [x] Name contains issue reference
    - [x] Has assignee
- [x] Documentation is updated (if required)
- [x] Tests are updated (if required)
- [x] Changes conform code style
- [x] CHANGELOG entry is added (if required)
- [x] FCM (final commit message) is posted
    - [x] and approved
- [x] [Review][l:2] is completed and changes are approved
- Before merge:
    - [x] Milestone is set
    - [x] PR's name and description are correct and up-to-date
    - [x] `Draft: ` prefix is removed
    - [x] All temporary labels are removed

[l:1]: https://help.github.com/en/articles/about-pull-requests#draft-pull-requests
[l:2]: https://help.github.com/en/articles/reviewing-changes-in-pull-requests

### CODE REVIEW

**Comment by ilslv** [comment]
## FCM

```
Implement `#[derive(GraphQLInterface)]` to use structs as GraphQL interfaces (#1026)

- support `#[graphql_interface]` on structs
```

File: juniper_codegen/src/graphql_interface/attr.rs [hunk 1]
```diff
@@ -243,22 +401,9 @@ fn err_default_impl_block<T, S: Spanned>(span: &S) -> Option<T> {
...
-
-/// Emits "no trait method receiver" [`syn::Error`] pointing to the given
-/// `span`.
-fn err_no_method_receiver<T, S: Spanned>(span: &S) -> Option<T> {
-    ERR.emit_custom(
-        span.span(),
-        "trait method should have a shared reference receiver `&self`",
-    );
+fn err_unnamed_field<T, S: Spanned>(span: &S) -> Option<T> {
+    ERR.emit_custom(span.span(), "expected named struct field");
```
Comment (tyranron) [inline juniper_codegen/src/graphql_interface/attr.rs:407]: I don't see this case covered in codegen failure tests. Please, add.

File: juniper_codegen/src/graphql_interface/derive.rs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        .unwrap_or_else(|| parse_quote! { () });
+
+    let enum_alias_ident = attr
+        .r#enum
+        .as_deref()
+        .cloned()
+        .unwrap_or_else(|| format_ident!("{}Value", struct_ident.to_string()));
+    let enum_ident = attr.r#enum.as_ref().map_or_else(
+        || format_ident!("{}ValueEnum", struct_ident.to_string()),
+        |c| format_ident!("{}Enum", c.inner().to_string()),
```
Comment (tyranron) [inline juniper_codegen/src/graphql_interface/derive.rs:95]: This code has become quite repetitive, so can be DRYed and factored out into a separate function for reusing.

File: juniper_codegen/src/graphql_interface/derive.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        ident: field_ident.clone(),
+        arguments: None,
+        has_receiver: false,
+        is_async: false,
+    })
+}
+
+/// Emits "expected named struct field" [`syn::Error`] pointing to the given
+/// `span`.
+fn err_unnamed_field<T, S: Spanned>(span: &S) -> Option<T> {
```
Comment (tyranron) [inline juniper_codegen/src/graphql_interface/derive.rs:173]: Meh... do not duplicate code. Just reuse the one from a neigbour module.

**Review by tyranron — CHANGES_REQUESTED** [review]
@ilslv overall good, but requires a little bit of polishing though.

**Comment by ilslv** [comment]
@tyranron CI build on nightly fails because of https://github.com/briansmith/ring/issues/1469 and from what I can tell this is a bug in recent nightly version and how it handles doc comments inside decl macros https://github.com/briansmith/ring/pull/1470

**Review by tyranron — APPROVED** [review]

**Pull request merged**